src/adapter/src/coord.rs

```diff
@@ -2835,9 +2882,65 @@ impl<S: Append + 'static> Coordinator<S> {
             config,
         }: CreateComputeInstanceReplicaPlan,
     ) -> Result<ExecuteResponse, AdapterError> {
+        // Choose default AZ if necessary
+        let config = match config {
+            ComputeInstanceReplicaConfig::Remote { addrs } => {
+                SerializedComputeInstanceReplicaConfig::Remote { addrs }
+            }
+            ComputeInstanceReplicaConfig::Managed {
+                size,
+                availability_zone,
+            } => {
+                let availability_zone = match availability_zone {
+                    Some(az) => {
+                        let azs = self.catalog.state().availability_zones();
+                        if !azs.contains(&az) {
+                            return Err(AdapterError::InvalidClusterReplicaAz {
+                                az: az.to_string(),
+                                expected: azs.to_vec(),
+                            });
+                        }
+                        az
+                    }
+                    None => {
+                        // Choose the least popular AZ among all replicas of this cluster as the default
+                        // if none was specified. If there is a tie for "least popular", pick the first one.
+                        // That is globally unbiased (for Materialize, not necessarily for this customer)
+                        // because we shuffle the AZs on boot.
+                        let instance = self.catalog.resolve_compute_instance(&of_cluster)?;
+                        let azs = self.catalog.state().availability_zones();
+                        let mut n_replicas_per_az = azs
+                            .iter()
+                            .map(|s| (s.as_str(), 0))
+                            .collect::<HashMap<_, _>>();
+                        for r in instance.replicas_by_id.values() {
+                            if let Some(az) = r.serialized_config.get_az() {
+                                let ct: &mut usize =
+                                    n_replicas_per_az.get_mut(az).expect("invalid AZ");
+                                *ct += 1
+                            }
+                        }
+                        let min = *n_replicas_per_az
+                            .values()
+                            .min()
+                            .expect("Must have at least one availability zone");
+                        let first_argmin = n_replicas_per_az
+                            .iter()
+                            .find_map(|(k, v)| (*v == min).then(|| k))
+                            .expect("Must have at least one availability zone");
+                        first_argmin.to_string()
+                    }
+                };
```

Comment:
Is there a way to extract a function to share this code between `CREATE CLUSTER` and `CREATE CLUSTER REPLICA`? Having two separate implementations worries me; they'll find a way to diverge.

Reply:
Well, the logic was different enough that it was hard to merge them, but now that you pointed out the logic for the multi-replica case is actually wrong, I think I'm just going to rip that out entirely and call into the single-replica logic in a loop. So, yeah, we can probably extract a function!